### PR TITLE
feature: add more signals - latency signal

### PR DESCRIPTION
**Pull request by liavweiss**
# Add Latency Signal for Model Selection Based on TPOT

## Summary

This PR implements a new **latency signal** that enables intelligent model selection based on real-time performance metrics, specifically Time Per Output Token (TPOT). This allows the router to automatically route requests to faster models when latency is a priority, improving user experience for real-time applications.

## Motivation

As discussed in [#1052](https://github.com/vllm-project/semantic-router/pull/1052), there was a need for:
**Latency signal based on TPOT** - Route to faster models for latency-sensitive scenarios

This implementation enables the router to make routing decisions based on real-time model performance metrics.

### TPOT (Time Per Output Token)

TPOT is calculated from each model response:
```
TPOT = completion_latency / completion_tokens
```

### TPOT Cache

- **In-memory cache** stores TPOT statistics per model
- **Exponential Moving Average (EMA)** with alpha=0.3 for smoothing
- **Automatic updates** from every model response
- **Thread-safe** using `sync.RWMutex`

### Latency Rules

Users define latency rules with `max_tpot` thresholds:

```yaml
signals:
  latency:
    - name: "low_latency"
      max_tpot: 0.05  # 50ms per token
      description: "For real-time chat applications requiring fast responses"
    - name: "medium_latency"
      max_tpot: 0.15  # 150ms per token
      description: "For standard applications with moderate latency tolerance"
```

### Signal Evaluation

1. **Model Collection**: Collects models from decisions that use latency signals
2. **TPOT Lookup**: Queries TPOT cache for each model
3. **Rule Matching**: Checks if model TPOT ≤ rule's `max_tpot` threshold
4. **Confidence Calculation**: Higher confidence for models with lower TPOT (further below threshold)

### Decision Integration

Latency signals can be combined with other signals (keyword, embedding, language, etc.) in decision rules:

```yaml
decisions:
  - name: "fast_english_route"
    rules:
      conditions:
        - type: "language"
          name: "en"
        - type: "latency"
          name: "low_latency"
      operator: "AND"
    model_refs:
      - model: "fast-model"
```

## Configuration Example

```yaml
signals:
  latency:
    - name: "low_latency"
      max_tpot: 0.05
      description: "For real-time chat applications"
    - name: "medium_latency"
      max_tpot: 0.15
      description: "For standard applications"

decisions:
  - name: "fast_route"
    rules:
      conditions:
        - type: "latency"
          name: "low_latency"
      operator: "AND"
    model_refs:
      - model: "fast-model-1"
      - model: "fast-model-2"
```

### Exponential Moving Average

- **Alpha = 0.3**: 30% weight to new observation, 70% to historical average
- **Rationale**: Smooths out temporary spikes while adapting to performance trends
- **Industry standard** for time-series smoothing

### Thread Safety

- `TPOTCache` uses `sync.RWMutex` for concurrent access
- Read locks for `GetTPOT()` (allows concurrent reads)
- Write locks for `UpdateTPOT()` (exclusive access)

### Model-Aware Design

Unlike query-based signals (keyword, embedding, language), latency is **model-aware**:
- Evaluates model performance, not query content
- Requires collecting models from relevant decisions
- TPOT data accumulates over time from actual responses

### Unit Tests
- ✅ TPOT cache operations
- ✅ Latency rule matching
- ✅ Edge cases (missing data, invalid values)
- ✅ Multiple models and rules
- ✅ All 147 tests passing

### Manual Testing
- ✅ Dashboard: Add/edit/delete latency signals
- ✅ Config validation
- ✅ Real-time TPOT updates
- ✅ Signal evaluation in routing decisions

## Breaking Changes

None. This is a new feature that doesn't affect existing functionality.

## Documentation

- ✅ Config template updated with examples
- ✅ Dashboard UI supports latency signal configuration
- ✅ Code comments explain TPOT cache and EMA algorithm

## Related Issues

- Addresses latency-based routing requirement from [#1052](https://github.com/vllm-project/semantic-router/pull/1052)
- **Part of:** #983

### CODE REVIEW

File: src/semantic-router/pkg/classification/latency_classifier.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+const TPOTAlpha = 0.3
+
+// TPOTCache stores recent TPOT values per model for latency signal evaluation
+type TPOTCache struct {
+	mu    sync.RWMutex
+	cache map[string]*ModelTPOTStats
+}
+
+// ModelTPOTStats stores TPOT statistics for a model
+type ModelTPOTStats struct {
```
Comment (rootfs) [inline src/semantic-router/pkg/classification/latency_classifier.go:23]: the absolute TPOT metrics (e.g. 10ms, 100ms) may not work out in production. In this case, a TPOT bucket is a better choice. Can you add a follow up PR to use TPOT bucket (e.g.  10% percentile, 50% percentile)?

**Review by rootfs — DISMISSED** [review]

File: src/semantic-router/pkg/config/config.go [hunk 2]
```diff
@@ -1687,6 +1692,21 @@ type LanguageRule struct {
...
+// LatencyRule defines a rule for latency-based signal classification
+// The latency classifier evaluates if available models meet TPOT (Time Per Output Token) requirements
+type LatencyRule struct {
+	// Name is the latency rule name that can be referenced in decision rules
+	Name string `yaml:"name"`
+
+	// MaxTPOT is the maximum acceptable TPOT (Time Per Output Token) in seconds
+	// Models with TPOT <= MaxTPOT will match this rule
+	// Example: 0.05 means 50ms per token
+	MaxTPOT float64 `yaml:"max_tpot"`
```
Comment (rootfs) [inline src/semantic-router/pkg/config/config.go:1792]: make this work for both absolute TPOT latency and latency percentile.

**Comment by Xunzhuo** [comment]
can you complete this pr like https://github.com/vllm-project/semantic-router/pull/1122 ?

**Comment by Xunzhuo** [comment]
the final target is to see the signal in the playground:

Clipboard_Screenshot_1768889039

File: src/semantic-router/pkg/classification/latency_classifier.go (hunk 1 above)
Comment (liavweiss) [reply src/semantic-router/pkg/classification/latency_classifier.go:23]: Yes, sure, I will.

**Comment by liavweiss** [comment]
Latency-signal
@Xunzhuo - I added the missing files.

I will add a follow-up PR to use the TPOT bucket ASAP
